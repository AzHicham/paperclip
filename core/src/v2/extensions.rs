use serde::{Deserialize, Deserializer, Serialize, Serializer};

use std::cmp::Ordering;
use std::collections::BTreeMap;
use std::ops::Deref;

/// Wrapper for `mime::MediaRange` to support `BTree{Set, Map}`.
#[derive(Debug, Clone)]
pub struct MediaRange(mime::MediaRange);

/// `x-encoder` and `x-decoder` global extension
/// for custom encoders and decoders.
#[derive(Debug, Default, Clone)]
pub struct Coders(pub BTreeMap<MediaRange, Coder>);

/// Represents the en/decoder for some MIME media range.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Coder {
    /// Path to the en/decoder function.
    pub coder_path: String,
    /// Path to the error type.
    pub error_path: String,
}

/* Common trait impls */

impl PartialEq for MediaRange {
    fn eq(&self, other: &MediaRange) -> bool {
        self.0.eq(&other.0)
    }
}

impl Eq for MediaRange {}

impl PartialOrd for MediaRange {
    fn partial_cmp(&self, other: &MediaRange) -> Option<Ordering> {
        Some(self.0.as_ref().cmp(other.0.as_ref()))
    }
}

impl Ord for MediaRange {
    fn cmp(&self, other: &MediaRange) -> Ordering {
        self.0.as_ref().cmp(other.0.as_ref())
    }
}

impl Serialize for MediaRange {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        self.0.serialize(serializer)
    }
}

impl<'de> Deserialize<'de> for MediaRange {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        Ok(MediaRange(mime::MediaRange::deserialize(deserializer)?))
    }
}

impl Deref for Coders {
    type Target = BTreeMap<MediaRange, Coder>;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl Serialize for Coders {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        self.0.serialize(serializer)
    }
}

impl<'de> Deserialize<'de> for Coders {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        Ok(Coders(BTreeMap::deserialize(deserializer)?))
    }
}
