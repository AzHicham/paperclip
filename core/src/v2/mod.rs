//! Core types and traits associated with the
//! [OpenAPI v2 specification](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md).

#[cfg(feature = "actix")]
mod actix;
mod extensions;
pub mod models;
mod resolver;
pub mod schema;

#[cfg(feature = "actix")]
pub use self::actix::{FutureWrapper, ResponderWrapper};

pub use self::models::{Api, DefaultSchema};
pub use self::schema::Schema;
pub use paperclip_macros::*;

use self::resolver::Resolver;
use crate::error::ValidationError;

impl<S: Schema> Api<S> {
    /// Consumes this API schema, resolves the references and returns
    /// the resolved schema.
    ///
    /// This walks recursively, collects the referenced schema objects,
    /// substitutes the referenced IDs with the pointer to schema objects
    /// and returns the resolved object or an error if it encountered one.
    pub fn resolve(self) -> Result<Api<S>, ValidationError> {
        let mut resolver = Resolver::from((self.definitions, self.paths));
        resolver.resolve()?;
        Ok(Api {
            swagger: self.swagger,
            definitions: resolver.defs,
            paths: resolver.paths,
            base_path: self.base_path,
            host: self.host,
            schemes: self.schemes,
            consumes: self.consumes,
            produces: self.produces,
            encoders: self.encoders,
            decoders: self.decoders,
            support_crates: self.support_crates,
        })
    }
}
