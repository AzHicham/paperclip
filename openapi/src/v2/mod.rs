//! Types and traits related to the [OpenAPI v2 spec](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md).
//!
//! # Detailed example
//!
//! To parse your v2 spec, you begin with transforming the schema into a
//! Rust struct. If your schema doesn't have custom properties, then you
//! can use the `DefaultSchema`.
//!
//! ```rust,no_run
//! use paperclip::v2::{self, Api, DefaultSchema, models::Version};
//!
//! use std::fs::File;
//!
//! let mut fd = File::open("my_spec.yaml").unwrap(); // yaml or json
//! let api: Api<DefaultSchema> = v2::from_reader(&mut fd).unwrap();
//! assert_eq!(api.swagger, Version::V2);
//! ```
//!
//! On the other hand, if your schema does have custom properties which you'd
//! like to parse, then use the `#[api_v2_schema]` proc macro.
//!
//! For example, let's take the [Kubernetes API spec][kube-spec]
//! which uses some custom thingmabobs. Let's say we're only interested in the
//! `x-kubernetes-patch-strategy` field for now.
//!
//! [kube-spec]: https://github.com/kubernetes/kubernetes/tree/afd928b8bc81cea385eba4c94558373df7aeae75/api/openapi-spec
//!
//! ```rust,no_run
//! #[macro_use] extern crate paperclip_macros;
//! #[macro_use] extern crate serde_derive; // NOTE: We're using serde for decoding stuff.
//!
//! use paperclip::v2::{self, Api};
//!
//! use std::fs::File;
//!
//! #[derive(Debug, Deserialize)]
//! #[serde(rename_all = "camelCase")]
//! enum PatchStrategy {
//!     Merge,
//!     RetainKeys,
//!     #[serde(rename = "merge,retainKeys")]
//!     MergeAndRetain,
//!     #[serde(other)]
//!     Other,
//! }
//!
//! #[api_v2_schema]
//! #[derive(Debug, Deserialize)]
//! struct K8sSchema {
//!     #[serde(rename = "x-kubernetes-patch-strategy")]
//!     patch_strategy: Option<PatchStrategy>,
//! }
//!
//! // K8sSchema now implements `Schema` trait.
//! let mut fd = File::open("k8s_spec.yaml").unwrap();
//! let api: Api<K8sSchema> = v2::from_reader(&mut fd).unwrap();
//! ```
//!
//! Next stop is to resolve this raw schema i.e., walk through the nodes,
//! find `$ref` fields and assign references to the corresponding definitions.
//!
//! ```rust,no_run
//! # use paperclip::v2::{self, Api, DefaultSchema};
//! # let api: Api<DefaultSchema> = v2::from_reader(&mut std::io::Cursor::new(vec![])).unwrap();
//!
//! let resolved = api.resolve().unwrap();
//! ```
//!
//! Now, if `codegen` feature is enabled (it is by default), we can use the
//! emitter to emit the API into some path.
//!
//! ```rust,no_run
//! # use paperclip::v2::{self, Api, DefaultSchema};
//! # let api: Api<DefaultSchema> = v2::from_reader(&mut std::io::Cursor::new(vec![])).unwrap();
//! use paperclip::v2::{DefaultEmitter, EmitterState, Emitter};
//!
//! let mut state = EmitterState::default();
//! state.working_dir = "/path/to/my/crate".into();
//! let emitter = DefaultEmitter::from(state);
//! emitter.generate(&api).unwrap(); // generate code!
//! ```

#[cfg(feature = "codegen")]
pub mod codegen;
pub mod im;
pub mod models;
mod resolver;

use self::models::{DataType, DataTypeFormat, SchemaRepr};
use self::resolver::Resolver;
use crate::error::PaperClipError;
use failure::Error;
use serde::Deserialize;

use std::collections::{BTreeMap, BTreeSet};
use std::io::{Read, Seek, SeekFrom};

#[cfg(feature = "codegen")]
pub use self::codegen::{DefaultEmitter, Emitter, EmitterState};
pub use self::models::{Api, DefaultSchema};

/// Deserialize the schema from the given reader. Currently, this only supports
/// JSON and YAML formats.
pub fn from_reader<R, S>(mut reader: R) -> Result<Api<S>, PaperClipError>
where
    R: Read + Seek,
    for<'de> S: Deserialize<'de>,
    S: Schema,
{
    let mut buf = [0; 1];
    reader.read_exact(&mut buf)?;
    reader.seek(SeekFrom::Start(0))?;

    if buf[0] == b'{' {
        // FIXME: Support whitespaces
        return Ok(serde_json::from_reader(reader)?);
    }

    Ok(serde_yaml::from_reader(reader)?)
}

/// Interface for the [`Schema`](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md#schemaObject) object.
///
/// This is only used for resolving the definitions.
///
/// **NOTE:** Don't implement this by yourself! Please use the `#[api_v2_schema]`
/// proc macro attribute instead.
pub trait Schema: Sized {
    /// Description for this schema, if any (`description` field).
    fn description(&self) -> Option<&str>;

    /// Reference to some other schema, if any (`$ref` field).
    fn reference(&self) -> Option<&str>;

    /// Data type of this schema, if any (`type` field).
    fn data_type(&self) -> Option<DataType>;

    /// Data type format used by this schema, if any (`format` field).
    fn format(&self) -> Option<&DataTypeFormat>;

    /// Schema for array definitions, if any (`items` field).
    fn items(&self) -> Option<&SchemaRepr<Self>>;

    /// Mutable access to the `items` field, if it exists.
    fn items_mut(&mut self) -> Option<&mut SchemaRepr<Self>>;

    /// Value schema for maps (`additional_properties` field).
    fn additional_properties(&self) -> Option<&SchemaRepr<Self>>;

    /// Mutable access to `additional_properties` field, if it's a map.
    fn additional_properties_mut(&mut self) -> Option<&mut SchemaRepr<Self>>;

    /// Map of names and schema for properties, if it's an object (`properties` field)
    fn properties(&self) -> Option<&BTreeMap<String, SchemaRepr<Self>>>;

    /// Mutable access to `properties` field.
    fn properties_mut(&mut self) -> Option<&mut BTreeMap<String, SchemaRepr<Self>>>;

    /// Returns the required properties (if any) for this object.
    fn required_properties(&self) -> Option<&BTreeSet<String>>;

    /// Set whether this definition is cyclic. This is done by the resolver.
    fn set_cyclic(&mut self, cyclic: bool);

    /// Returns whether this definition is cyclic.
    ///
    /// **NOTE:** This is not part of the schema object, but it's
    /// set by the resolver using `set_cyclic` for codegen.
    fn is_cyclic(&self) -> bool;

    /// Name of this schema, if any.
    ///
    /// **NOTE:** This is not part of the schema object, but it's
    /// set by the resolver using `set_name` for codegen.
    fn name(&self) -> Option<&str>;

    /// Sets the name for this schema. This is done by the resolver.
    fn set_name(&mut self, name: &str);
}

impl<S: Schema> Api<S> {
    /// Consumes this API schema, resolves the references and returns
    /// the resolved schema.
    ///
    /// This walks recursively, collects the referenced schema objects,
    /// substitutes the referenced IDs with the pointer to schema objects
    /// and returns the resolved object or an error if it encountered one.
    pub fn resolve(self) -> Result<Api<S>, Error> {
        let mut resolver = Resolver::from((self.definitions, self.paths));
        resolver.resolve()?;
        Ok(Api {
            swagger: self.swagger,
            definitions: resolver.defs,
            paths: resolver.paths,
        })
    }
}
