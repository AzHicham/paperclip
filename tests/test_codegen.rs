#[macro_use]
extern crate lazy_static;

use paperclip::v2::{
    self,
    codegen::{CrateMeta, DefaultEmitter, Emitter, EmitterState},
    models::{Api, DefaultSchema},
};

use std::fs::File;
use std::io::Read;

lazy_static! {
    static ref ROOT: String = String::from(env!("CARGO_MANIFEST_DIR"));
    static ref SCHEMA: Api<DefaultSchema> = {
        let fd = File::open(ROOT.clone() + "/tests/pet-v2.yaml").expect("file?");
        let raw: Api<DefaultSchema> = v2::from_reader(fd).expect("deserializing spec");
        raw.resolve().expect("resolution")
    };
    static ref CODEGEN: () = {
        let mut state = EmitterState::default();
        state.working_dir = (&*ROOT).into();
        state.working_dir.push("tests/test_pet");
        let mut meta = CrateMeta::default();
        meta.authors = Some(vec!["Me <me@example.com>".into()]);
        state.set_meta(meta);

        let emitter = DefaultEmitter::from(state);
        emitter.generate(&SCHEMA).expect("codegen");
    };
    static ref CLI_CODEGEN: () = {
        let _ = &*CODEGEN;
        let mut state = EmitterState::default();
        state.working_dir = (&*ROOT).into();
        state.working_dir.push("tests/test_pet/cli");
        let mut meta = CrateMeta::default();
        meta.authors = Some(vec!["Me <me@example.com>".into()]);
        meta.is_cli = true;
        state.set_meta(meta);

        let emitter = DefaultEmitter::from(state);
        emitter.generate(&SCHEMA).expect("codegen");
    };
}

fn assert_file_contains_content_at(path: &str, matching_content: &str, index: Option<usize>) {
    let _ = &*CODEGEN;

    let mut contents = String::new();
    let mut fd = File::open(path).expect("missing file");
    fd.read_to_string(&mut contents).expect("reading file");

    let result = contents.find(matching_content);
    if let Some(idx) = index {
        assert_eq!(result, Some(idx));
    } else {
        assert!(result.is_some());
    }
}

#[test]
fn test_lib_creation() {
    assert_file_contains_content_at(
        &(ROOT.clone() + "/tests/test_pet/lib.rs"),
        "
#[macro_use]
extern crate failure;
#[macro_use]
extern crate serde;

pub mod category {
    include!(\"./category.rs\");
}

pub mod order {
    include!(\"./order.rs\");
}

pub mod pet {
    include!(\"./pet.rs\");
}

pub mod tag {
    include!(\"./tag.rs\");
}

pub mod client {
    use futures::{Future, future};
",
        Some(0),
    );
}

#[test]
fn test_manifest_creation() {
    assert_file_contains_content_at(
        &(ROOT.clone() + "/tests/test_pet/Cargo.toml"),
        "[package]
name = \"test_pet\"
version = \"0.1.0\"
authors = [\"Me <me@example.com>\"]
edition = \"2018\"

[lib]
path = \"lib.rs\"

[dependencies]
failure = \"0.1\"
futures = \"0.1\"
parking_lot = \"0.8\"
reqwest = \"0.9\"
serde = \"1.0\"
",
        Some(0),
    );
}

#[test]
fn test_overridden_path() {
    // We've specified `host` and `basePath` in our spec, so that should be used in place of the placeholder.
    assert_file_contains_content_at(
        &(ROOT.clone() + "/tests/test_pet/lib.rs"),
        "
    impl ApiClient for reqwest::r#async::Client {
        #[inline]
        fn request_builder(&self, method: reqwest::Method, rel_path: &str) -> reqwest::r#async::RequestBuilder {
            let mut u = String::from(\"https://pets.com:8888/api\");
            u.push_str(rel_path.trim_start_matches('/'));
            self.request(method, &u)
        }
",
        Some(1630),
    );
}

#[test]
fn test_header_parameters() {
    assert_file_contains_content_at(
        &(ROOT.clone() + "/tests/test_pet/pet.rs"),
        "
impl<XAuth, Id, Name> PetPostBuilder<XAuth, Id, Name> {
    #[inline]
    pub fn x_auth(mut self, value: impl Into<String>) -> PetPostBuilder<crate::generics::XAuthExists, Id, Name> {
        self.inner.param_x_auth = Some(value.into());
        unsafe { std::mem::transmute(self) }
    }

    #[inline]
    pub fn x_pet_id(mut self, value: impl Into<i64>) -> Self {
        self.inner.param_x_pet_id = Some(value.into());
        self
    }
",
        Some(3938),
    );

    assert_file_contains_content_at(
        &(ROOT.clone() + "/tests/test_pet/pet.rs"),
        "
impl crate::client::Sendable for PetPostBuilder<crate::generics::XAuthExists, crate::generics::IdExists, crate::generics::NameExists> {
    type Output = crate::pet::Pet;

    const METHOD: reqwest::Method = reqwest::Method::POST;

    fn rel_path(&self) -> std::borrow::Cow<'static, str> {
        \"/pets\".into()
    }

    fn modify(&self, req: reqwest::r#async::RequestBuilder) -> reqwest::r#async::RequestBuilder {
        let mut req = req;
        req = req.header(\"X-Auth\", self.inner.param_x_auth.as_ref().map(std::string::ToString::to_string).expect(\"missing parameter x_auth?\"));
        if let Some(v) = self.inner.param_x_pet_id.as_ref().map(std::string::ToString::to_string) {
            req = req.header(\"X-Pet-ID\", v);
        }

        req
        .json(&self.inner.body)
    }
}
",
        Some(5411),
    );
}

#[test]
fn test_array_response() {
    // If an operation returns an array of objects, then we bind that
    // operation to that object and do `Sendable<Output<Vec<Object>>>`.
    assert_file_contains_content_at(
        &(ROOT.clone() + "/tests/test_pet/pet.rs"),
        "
/// Builder created by [`Pet::list_pets`](./struct.Pet.html#method.list_pets) method for a `GET` operation associated with `Pet`.
#[derive(Debug, Clone)]
pub struct PetGetBuilder;


impl crate::client::Sendable for PetGetBuilder {
    type Output = Vec<Pet>;

    const METHOD: reqwest::Method = reqwest::Method::GET;

    fn rel_path(&self) -> std::borrow::Cow<'static, str> {
        \"/pets\".into()
    }
}
",
        Some(2978),
    );
}

#[test]
fn test_operation_with_payload_no_arguments() {
    let _ = &*CLI_CODEGEN;
    // An operation with no arguments should enforce payload if it needs one.
    assert_file_contains_content_at(
        &(ROOT.clone() + "/tests/test_pet/cli/app.yaml"),
        "
  - add-pet:
      about: \"Add a new pet to the store\"
      args:
        - payload:
            long: payload
            help: \"Path to payload (schema: Pet) or pass '-' for stdin\"
            takes_value: true
            required: true
",
        None,
    );
}

#[test]
fn test_anonymous_object_definitions() {
    let _ = &*CLI_CODEGEN;
    // An object "can" define objects in its schema without referencing
    // them from known definitions. In that case, we autogenerate stuff.
    assert_file_contains_content_at(
        &(ROOT.clone() + "/tests/test_pet/order.rs"),
        "#[derive(Debug, Default, Clone, Deserialize, Serialize)]
pub struct Order {
    pub address: Option<OrderAddress>,
    pub id: Option<i64>,
    pub list: Option<Vec<OrderList>>,
}
#[derive(Debug, Default, Clone, Deserialize, Serialize)]
pub struct OrderAddress {
    pub code: Option<String>,
    pub line1: Option<String>,
    pub line2: Option<String>,
    pub name: Option<String>,
}
#[derive(Debug, Default, Clone, Deserialize, Serialize)]
pub struct OrderList {
    #[serde(rename = \"petId\")]
    pub pet_id: Option<i64>,
    pub quantity: Option<i64>,
}

impl Order {
    /// Create a builder for this object.
    #[inline]
    pub fn builder() -> OrderBuilder {
        OrderBuilder {
            body: Default::default(),
        }
    }
}

impl Into<Order> for OrderBuilder {
    fn into(self) -> Order {
        self.body
    }
}

/// Builder for [`Order`](./struct.Order.html) object.
#[derive(Debug, Clone)]
pub struct OrderBuilder {
    body: self::Order,
}

impl OrderBuilder {
    #[inline]
    pub fn address(mut self, value: impl Into<OrderAddress>) -> Self {
        self.body.address = Some(value.into());
        self
    }

    #[inline]
    pub fn id(mut self, value: impl Into<i64>) -> Self {
        self.body.id = Some(value.into());
        self
    }

    #[inline]
    pub fn list(mut self, value: impl Iterator<Item = impl Into<OrderList>>) -> Self {
        self.body.list = Some(value.map(|value| value.into()).collect::<Vec<_>>());
        self
    }
}

impl OrderAddress {
    /// Create a builder for this object.
    #[inline]
    pub fn builder() -> OrderAddressBuilder {
        OrderAddressBuilder {
            body: Default::default(),
        }
    }
}

impl Into<OrderAddress> for OrderAddressBuilder {
    fn into(self) -> OrderAddress {
        self.body
    }
}

/// Builder for [`OrderAddress`](./struct.OrderAddress.html) object.
#[derive(Debug, Clone)]
pub struct OrderAddressBuilder {
    body: self::OrderAddress,
}

impl OrderAddressBuilder {
    #[inline]
    pub fn code(mut self, value: impl Into<String>) -> Self {
        self.body.code = Some(value.into());
        self
    }

    #[inline]
    pub fn line1(mut self, value: impl Into<String>) -> Self {
        self.body.line1 = Some(value.into());
        self
    }

    #[inline]
    pub fn line2(mut self, value: impl Into<String>) -> Self {
        self.body.line2 = Some(value.into());
        self
    }

    #[inline]
    pub fn name(mut self, value: impl Into<String>) -> Self {
        self.body.name = Some(value.into());
        self
    }
}

impl OrderList {
    /// Create a builder for this object.
    #[inline]
    pub fn builder() -> OrderListBuilder {
        OrderListBuilder {
            body: Default::default(),
        }
    }
}

impl Into<OrderList> for OrderListBuilder {
    fn into(self) -> OrderList {
        self.body
    }
}

/// Builder for [`OrderList`](./struct.OrderList.html) object.
#[derive(Debug, Clone)]
pub struct OrderListBuilder {
    body: self::OrderList,
}

impl OrderListBuilder {
    #[inline]
    pub fn pet_id(mut self, value: impl Into<i64>) -> Self {
        self.body.pet_id = Some(value.into());
        self
    }

    #[inline]
    pub fn quantity(mut self, value: impl Into<i64>) -> Self {
        self.body.quantity = Some(value.into());
        self
    }
}
",
        Some(0),
    );
}
