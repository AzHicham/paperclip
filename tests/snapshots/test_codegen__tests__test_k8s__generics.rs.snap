---
source: tests/test_codegen.rs
expression: data
---

pub struct MissingAcceptedNames;
pub struct AcceptedNamesExists;
pub struct MissingAddress;
pub struct AddressExists;
pub struct MissingAllowed;
pub struct AllowedExists;
pub struct MissingAllowedRuntimeClassNames;
pub struct AllowedRuntimeClassNamesExists;
pub struct MissingApiGroup;
pub struct ApiGroupExists;
pub struct MissingApiVersion;
pub struct ApiVersionExists;
pub struct MissingArchitecture;
pub struct ArchitectureExists;
pub struct MissingAttached;
pub struct AttachedExists;
pub struct MissingAttacher;
pub struct AttacherExists;
pub struct MissingBackend;
pub struct BackendExists;
pub struct MissingBootId;
pub struct BootIdExists;
pub struct MissingBuildDate;
pub struct BuildDateExists;
pub struct MissingCidr;
pub struct CidrExists;
pub struct MissingClaimName;
pub struct ClaimNameExists;
pub struct MissingClientCidr;
pub struct ClientCidrExists;
pub struct MissingClientConfig;
pub struct ClientConfigExists;
pub struct MissingCompiler;
pub struct CompilerExists;
pub struct MissingConditionType;
pub struct ConditionTypeExists;
pub struct MissingConditions;
pub struct ConditionsExists;
pub struct MissingContainerPort;
pub struct ContainerPortExists;
pub struct MissingContainerRuntimeVersion;
pub struct ContainerRuntimeVersionExists;
pub struct MissingContainers;
pub struct ContainersExists;
pub struct MissingCount;
pub struct CountExists;
pub struct MissingCurrent;
pub struct CurrentExists;
pub struct MissingCurrentAverageValue;
pub struct CurrentAverageValueExists;
pub struct MissingCurrentHealthy;
pub struct CurrentHealthyExists;
pub struct MissingCurrentNumberScheduled;
pub struct CurrentNumberScheduledExists;
pub struct MissingCurrentReplicas;
pub struct CurrentReplicasExists;
pub struct MissingCurrentValue;
pub struct CurrentValueExists;
pub struct MissingDescribedObject;
pub struct DescribedObjectExists;
pub struct MissingDesiredHealthy;
pub struct DesiredHealthyExists;
pub struct MissingDesiredNumberScheduled;
pub struct DesiredNumberScheduledExists;
pub struct MissingDesiredReplicas;
pub struct DesiredReplicasExists;
pub struct MissingDevicePath;
pub struct DevicePathExists;
pub struct MissingDiskName;
pub struct DiskNameExists;
pub struct MissingDiskUri;
pub struct DiskUriExists;
pub struct MissingDisruptionsAllowed;
pub struct DisruptionsAllowedExists;
pub struct MissingDriver;
pub struct DriverExists;
pub struct MissingDrivers;
pub struct DriversExists;
pub struct MissingEffect;
pub struct EffectExists;
pub struct MissingEndpoints;
pub struct EndpointsExists;
pub struct MissingEventTime;
pub struct EventTimeExists;
pub struct MissingExitCode;
pub struct ExitCodeExists;
pub struct MissingExpectedPods;
pub struct ExpectedPodsExists;
pub struct MissingFieldPath;
pub struct FieldPathExists;
pub struct MissingFsGroup;
pub struct FsGroupExists;
pub struct MissingGateway;
pub struct GatewayExists;
pub struct MissingGitCommit;
pub struct GitCommitExists;
pub struct MissingGitTreeState;
pub struct GitTreeStateExists;
pub struct MissingGitVersion;
pub struct GitVersionExists;
pub struct MissingGoVersion;
pub struct GoVersionExists;
pub struct MissingGroup;
pub struct GroupExists;
pub struct MissingGroupPriorityMinimum;
pub struct GroupPriorityMinimumExists;
pub struct MissingGroupVersion;
pub struct GroupVersionExists;
pub struct MissingGroups;
pub struct GroupsExists;
pub struct MissingHandler;
pub struct HandlerExists;
pub struct MissingImage;
pub struct ImageExists;
pub struct MissingImageId;
pub struct ImageIdExists;
pub struct MissingIncomplete;
pub struct IncompleteExists;
pub struct MissingInvolvedObject;
pub struct InvolvedObjectExists;
pub struct MissingIp;
pub struct IpExists;
pub struct MissingIqn;
pub struct IqnExists;
pub struct MissingItems;
pub struct ItemsExists;
pub struct MissingJobTemplate;
pub struct JobTemplateExists;
pub struct MissingJsonPath;
pub struct JsonPathExists;
pub struct MissingKernelVersion;
pub struct KernelVersionExists;
pub struct MissingKey;
pub struct KeyExists;
pub struct MissingKind;
pub struct KindExists;
pub struct MissingKubeProxyVersion;
pub struct KubeProxyVersionExists;
pub struct MissingKubeletConfigKey;
pub struct KubeletConfigKeyExists;
pub struct MissingKubeletVersion;
pub struct KubeletVersionExists;
pub struct MissingLastObservedTime;
pub struct LastObservedTimeExists;
pub struct MissingLevel;
pub struct LevelExists;
pub struct MissingLimits;
pub struct LimitsExists;
pub struct MissingLun;
pub struct LunExists;
pub struct MissingMachineId;
pub struct MachineIdExists;
pub struct MissingMajor;
pub struct MajorExists;
pub struct MissingMax;
pub struct MaxExists;
pub struct MissingMaxReplicas;
pub struct MaxReplicasExists;
pub struct MissingMetadata;
pub struct MetadataExists;
pub struct MissingMetric;
pub struct MetricExists;
pub struct MissingMetricName;
pub struct MetricNameExists;
pub struct MissingMin;
pub struct MinExists;
pub struct MissingMinor;
pub struct MinorExists;
pub struct MissingMonitors;
pub struct MonitorsExists;
pub struct MissingMountPath;
pub struct MountPathExists;
pub struct MissingName;
pub struct NameExists;
pub struct MissingNames;
pub struct NamesExists;
pub struct MissingNamespace;
pub struct NamespaceExists;
pub struct MissingNamespaced;
pub struct NamespacedExists;
pub struct MissingNodeId;
pub struct NodeIdExists;
pub struct MissingNodeName;
pub struct NodeNameExists;
pub struct MissingNodeSelectorTerms;
pub struct NodeSelectorTermsExists;
pub struct MissingNonResourceRules;
pub struct NonResourceRulesExists;
pub struct MissingNumberMisscheduled;
pub struct NumberMisscheduledExists;
pub struct MissingNumberReady;
pub struct NumberReadyExists;
pub struct MissingObject;
pub struct ObjectExists;
pub struct MissingOperatingSystem;
pub struct OperatingSystemExists;
pub struct MissingOperator;
pub struct OperatorExists;
pub struct MissingOsImage;
pub struct OsImageExists;
pub struct MissingPath;
pub struct PathExists;
pub struct MissingPaths;
pub struct PathsExists;
pub struct MissingPdId;
pub struct PdIdExists;
pub struct MissingPdName;
pub struct PdNameExists;
pub struct MissingPending;
pub struct PendingExists;
pub struct MissingPlatform;
pub struct PlatformExists;
pub struct MissingPlural;
pub struct PluralExists;
pub struct MissingPodAffinityTerm;
pub struct PodAffinityTermExists;
pub struct MissingPodSelector;
pub struct PodSelectorExists;
pub struct MissingPolicy;
pub struct PolicyExists;
pub struct MissingPort;
pub struct PortExists;
pub struct MissingPreference;
pub struct PreferenceExists;
pub struct MissingProvisioner;
pub struct ProvisionerExists;
pub struct MissingRaw;
pub struct RawExists;
pub struct MissingReady;
pub struct ReadyExists;
pub struct MissingRegistry;
pub struct RegistryExists;
pub struct MissingReplicas;
pub struct ReplicasExists;
pub struct MissingRepository;
pub struct RepositoryExists;
pub struct MissingRequest;
pub struct RequestExists;
pub struct MissingResource;
pub struct ResourceExists;
pub struct MissingResourceRules;
pub struct ResourceRulesExists;
pub struct MissingResources;
pub struct ResourcesExists;
pub struct MissingRestartCount;
pub struct RestartCountExists;
pub struct MissingRevision;
pub struct RevisionExists;
pub struct MissingRoleRef;
pub struct RoleRefExists;
pub struct MissingRollbackTo;
pub struct RollbackToExists;
pub struct MissingRule;
pub struct RuleExists;
pub struct MissingRunAsUser;
pub struct RunAsUserExists;
pub struct MissingRuntimeHandler;
pub struct RuntimeHandlerExists;
pub struct MissingScaleTargetRef;
pub struct ScaleTargetRefExists;
pub struct MissingSchedule;
pub struct ScheduleExists;
pub struct MissingScope;
pub struct ScopeExists;
pub struct MissingScopeName;
pub struct ScopeNameExists;
pub struct MissingSeLinux;
pub struct SeLinuxExists;
pub struct MissingSecretName;
pub struct SecretNameExists;
pub struct MissingSecretRef;
pub struct SecretRefExists;
pub struct MissingSelector;
pub struct SelectorExists;
pub struct MissingServed;
pub struct ServedExists;
pub struct MissingServer;
pub struct ServerExists;
pub struct MissingServerAddress;
pub struct ServerAddressExists;
pub struct MissingServerAddressByClientCidRs;
pub struct ServerAddressByClientCidRsExists;
pub struct MissingService;
pub struct ServiceExists;
pub struct MissingServiceName;
pub struct ServiceNameExists;
pub struct MissingServicePort;
pub struct ServicePortExists;
pub struct MissingShareName;
pub struct ShareNameExists;
pub struct MissingSingularName;
pub struct SingularNameExists;
pub struct MissingSource;
pub struct SourceExists;
pub struct MissingSources;
pub struct SourcesExists;
pub struct MissingSpec;
pub struct SpecExists;
pub struct MissingSpecReplicasPath;
pub struct SpecReplicasPathExists;
pub struct MissingState;
pub struct StateExists;
pub struct MissingStatus;
pub struct StatusExists;
pub struct MissingStatusReplicasPath;
pub struct StatusReplicasPathExists;
pub struct MissingStorage;
pub struct StorageExists;
pub struct MissingStoredVersions;
pub struct StoredVersionsExists;
pub struct MissingStrategy;
pub struct StrategyExists;
pub struct MissingSupplementalGroups;
pub struct SupplementalGroupsExists;
pub struct MissingSystem;
pub struct SystemExists;
pub struct MissingSystemUuid;
pub struct SystemUuidExists;
pub struct MissingTarget;
pub struct TargetExists;
pub struct MissingTargetAverageValue;
pub struct TargetAverageValueExists;
pub struct MissingTargetPortal;
pub struct TargetPortalExists;
pub struct MissingTargetValue;
pub struct TargetValueExists;
pub struct MissingTemplate;
pub struct TemplateExists;
pub struct MissingTopologyKey;
pub struct TopologyKeyExists;
pub struct MissingType;
pub struct TypeExists;
pub struct MissingUid;
pub struct UidExists;
pub struct MissingValue;
pub struct ValueExists;
pub struct MissingValues;
pub struct ValuesExists;
pub struct MissingVerbs;
pub struct VerbsExists;
pub struct MissingVersion;
pub struct VersionExists;
pub struct MissingVersionPriority;
pub struct VersionPriorityExists;
pub struct MissingVersions;
pub struct VersionsExists;
pub struct MissingVolume;
pub struct VolumeExists;
pub struct MissingVolumeHandle;
pub struct VolumeHandleExists;
pub struct MissingVolumeId;
pub struct VolumeIdExists;
pub struct MissingVolumePath;
pub struct VolumePathExists;
pub struct MissingWebhook;
pub struct WebhookExists;
pub struct MissingWeight;
pub struct WeightExists;

